### DISCUSSION TITLE
Move essential examples back to main repo and test them in CI

### DISCUSSION THREAD

**Discussion by EwoutH** [mesa/mesa#1929] [General]
All mesa examples were moved to [mesa-examples](https://github.com/projectmesa/mesa-examples) in #1529. While unfortunately that PR doesn't provide or link to motivation for that move, I would propose moving at least some of them - and maybe all of them - back to the main Mesa repo.

This has four distinct advantages:
- The examples will be more visible and easier to install for users.
- The examples will be part of tagged releases, ensuring that examples also follow versioning and are in sync with the Mesa functionality for each release.
- The examples can be (easier) tested in CI, to detect (breaking) changes.

To make that work, when new functionality is included in a PR, we could require the essential examples to be updated as well. This way they will always be up to date and show if any (breaking) changes need to be made.

**Comment by wang-boyu** [comment]
If I recall correctly, one of the motivations is to have separate repos for devs and users. Devs/maintainers will be mostly working on the source code repo. Users, on the other hand, will be mainly using readthedocs site and the examples repo, without looking into Mesa's source code.

Previously I suggested to have a dedicated website for example models: https://github.com/projectmesa/mesa-examples/discussions/61. It would also be interesting to have example models from community contributions: https://github.com/projectmesa/mesa-examples/issues/11#issuecomment-1789933554. Separating example models from source code repo would make such efforts easier to maintain, IMHO.

**Reply by EwoutH** [reply]
Thanks for the context!

I'm thinking in a direction to have a few core examples included in the main repo (5 to 10), on which version control and CI is ran. Then we can have a larger "best effort" repo with a lot more examples.

A dedicated website would be definitely amazing, especially if they're interactive (like [NetLogo Web](https://www.netlogoweb.org/launch)).

**Reply by Corvince** [reply]
Yes i think it was one of the original ideas to have some core examples that also get some more love with updating in the main repo. And the examples repo to have more of a model zoo where people can more easily contribute models. 

One thing to note is that we should also be able to run the examples as part of our ci even when they are on a different repo. First step would be to have a test in the examples repo, I already submitted a PR for that. Then we just have to clone the repo and run the test.

But I believe it would still be nice to have some examples in the main repo to more easily run them locally as part of the test runner and also to possibly highlight them further.

**Reply by EwoutH** [reply]
How feasible do you both think something like NetLogo Web but for Mesa would be?

**Reply by wang-boyu** [reply]
It has become more feasible since the support for Jupyter notebooks - the least we can do is to convert examples into notebooks, have a static webpage with all the notebook, and let users run by themselves on Binder or Colab (example: https://juliaclimate.github.io/Notebooks/ mentioned in https://github.com/projectmesa/mesa-examples/discussions/61#discussioncomment-7080517)

Another possibility is to deploy all examples on some server, example: https://simrisk.pulcloud.io/riskyfood mentioned in https://github.com/projectmesa/mesa/issues/1906#issue-2043954676 and https://github.com/projectmesa/mesa/issues/1800

This is something that I am hoping for, especially if we can have the community to contribute their models (mentioned in https://github.com/projectmesa/mesa-examples/issues/11#issuecomment-1789933554). I also think that it would be nice for some users to only refer to readthedocs site and an example webpage to learn Mesa, without having to visit GitHub at all.

**Reply by Corvince** [reply]
With you mentioning notebooks - There is also [JupyterLite](https://jupyter.org/try-jupyter/lab/index.html), which allows you to host notebooks statically, but they can directly be run client-side (locally). I think this is super exciting. I already played around with it a bit and want to deploy a test site soon with some of the mesa model notebooks. We then just need to find a way to update them regularly, maybe this can be done through Github Actions.

But the current downside is that solara is not yet supported by Pyodide (which makes JupyterLite possible). But its is on the radar of the solara team and was originally planned for end of last year. If everything falls into place we can then even facilitate a way for users to host their own models statically (which is a lot easier and free on lots of sites, e.g. GitHub Pages)

**Reply by EwoutH** [reply]
Sounds amazing, I can support on the CI side.

There is an issue we can track: https://github.com/widgetti/solara/issues/449

**Comment by EwoutH** [comment]
Let's try to get a list of models that we want in the main repo (and why).

I would suggest:
- [Boltzmann wealth model](https://github.com/projectmesa/mesa-examples/tree/main/examples/boltzmann_wealth_model). Very easy, very simple, good to start with.
  - RandomActivation, MultiGrid
- [Boid flockers](https://github.com/projectmesa/mesa-examples/tree/main/examples/boid_flockers). Good demo of agent movements and the continuous space
  - RandomActivation, ContinuousSpace
- [Conway's Game Of Life](https://github.com/projectmesa/mesa-examples/tree/main/examples/conways_game_of_life). Classic, shows nice emergent properties. Candidate for PropertyLayer implementation.
  - SimultaneousActivation, SingleGrid
- [Wolf-Sheep Predation](https://github.com/projectmesa/mesa-examples/tree/main/examples/wolf_sheep). Multiple agent types, agent creation and deletion. Candidate for PropertyLayer methods.
  - RandomActivationByTypeFiltered, MultiGrid
- [Virus on a Network](https://github.com/projectmesa/mesa-examples/tree/main/examples/virus_on_network). Mainly to test the NetworkGrid.
  - RandomActivation, NetworkGrid
- [Hexagonal GoL](https://github.com/projectmesa/mesa-examples/tree/main/examples/hex_snowflake). A (the only) demo of HexGrid.
  - SimultaneousActivation, HexGrid

**Comment by rht** [comment]
Prior thread: https://github.com/projectmesa/mesa/issues/134. Another reason for having all the examples in 1 repo is that you (both developers and users) know exactly there is 1 repo where all the examples reside, and without having consciously remember which example is in the examples repo, or which example is in the core Mesa repo. Also, there is an issue where most discussion threads, I have pointed to the mesa-examples repo, and to move some of them back would break the links.

**Comment by EwoutH** [comment]
In popular demand, I will develop a benchmark script, that can both be run locally and in CI that reports back on each GitHub PR. Therefor it needs some model in the repo (to be able to run it locally).
1. Do we want the models purely for the benchmarks, or also as examples?
2. Which models do we want for benchmarking?

**Reply by rht** [reply]
> Which models do we want for benchmarking?

The ones used in https://github.com/JuliaDynamics/ABM_Framework_Comparisons.

**Reply by EwoutH** [reply]
- Flocking
- Schelling
- WolfSheep

Any others?

**Reply by rht** [reply]
https://github.com/projectmesa/mesa/discussions/1610#discussioncomment-6155199 has analysis on what is the benchmark focus by each examples.

**Reply by Corvince** [reply]
Thanks for doing this @EwoutH ! I think the models from the ABM comparison are a good start, maybe really just concentrate on them first. That said they are primarily selected because they are present in all the libraries compared. Otherwise we would have to look for models that use as much of the provided functionality in mesa.

**Reply by quaquel** [reply]
Indeed, let's start with the ABM comparison models first. This will give a helpful structure and workflow. We can always add other models that test specific parts not covered by the current models (e.g., virus on a network for NetworkGrid).

**Reply by EwoutH** [reply]
Would you prefer the model versions from [mesa-examples](https://github.com/projectmesa/mesa-examples) or [ABMFrameworksComparison](https://github.com/JuliaDynamics/ABMFrameworksComparison)?

**Reply by quaquel** [reply]
Are there significant differences? It would be good to use the same as the comparison and also push any changes in light of new API's to there. Examples, in my view, are also often used to illustrate functionality and need not be the most performant way of doing things.

**Reply by EwoutH** [reply]
Agreed, will start with the current ABMFrameworksComparison and we can update from there.

(already thinking about fixed seeds, multiple runs per seed, statistical comparisons. why can't I ever do things half?)

**Reply by quaquel** [reply]
I wouldn't worry about multiple runs per seed and statistical comparisons. It is just a simple quick test, not a detailed comparison that needs to get published.

**Reply by Corvince** [reply]
I mean the comparison is probably a really good start, because I think they have already considered some of the things you mentioned. They are also very respectful and helpful, maybe you can also funnel back some ideas to them

**Reply by EwoutH** [reply]
Nailed the benchmarking itself. Fixed seeds, multiple replications per seeds, only takes the fastest of n replications (in performance benchmarking you focus on fastest, since negative outliers are fundamentally limited). Multiple models, multiple configurations per model. Measures model initialisation and run time separately.

Working branch: https://github.com/EwoutH/mesa/tree/benchmarks-ci

The difficulty now is that you need something to compare it to, and for that you will need some plumbing with git and GitHub Actions. Ideally I get the comparisons working both locally and in CI, but locally is a bit more difficult, because everyone's git environment is different, so you can't just checkout a branch as easily as when using a stable CI environment.

Probably will resort to saving the old results in a file, then saving the new results in a file, and then comparing both files.

What helps it's only for developers so it doesn't need to be obviously easy to use.

So all surmountable challenges, will likely have a PR tomorrow.

**Reply by quaquel** [reply]
This. is great and beyond what I had in mind.

Perhaps try not to overcomplicate things at the moment. Locally, I can switch between branches and rerun the script. This is 2 seconds of work, so why try to automate all of this? For comparing results, it would be convenient if the user could pass a directory where the results can be stored. This cannot be within the directory of the benchmarks because that is under git control. Perhaps autogenerate the name of the CSV file based on the name of the branch (or timestamp it). I can then run both branches locally, import both csv files into a notebook, and quickly compare them. 

Complementary, you might want to have some command line support where you print key statistics to stdout, switch branches, run again and you can see your results.

**Reply by EwoutH** [reply]
Thanks for your insights. Just realized that I just could add results files to the`.gitignore`, I think that would be quite elegant.

**Reply by quaquel** [reply]
what happens with files that are under .gitignore if you switch branch?

**Reply by EwoutH** [reply]
They are not tracked in git at all, so they remain unchanged.

**Reply by quaquel** [reply]
I thought so. Thanks for confirming. Then, that is the simple fix.

**Reply by EwoutH** [reply]
Stuff working locally. I think it can quite reliably detect changes +- 5% in model init and runtimes (depending on system of course).

Next up CI!

**Reply by quaquel** [reply]
Good stuff! Will be great to test in conjunction with @Corvince's work on an alternative grid implementation. Is the code online up to date?

**Reply by EwoutH** [reply]
Yes
